File: leetcode/tree/lowest_common_ancestor_bs.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	var res *TreeNode
	helper(root, p, q, &res)
	return res
}

func helper(root, p, q *TreeNode, res **TreeNode) bool {
	if root == nil {
		return false
	}
	left := helper(root.Left, p, q, res)
	right := helper(root.Right, p, q, res)
	mid := root == p || root == q
	if (mid && left) || (mid && right) || (right && left) {
		*res = root
	}
	return mid || left || right
}

func main() {
	root := &TreeNode{6, nil, nil}
	root.Left = &TreeNode{2, nil, nil}
	root.Right = &TreeNode{8, nil, nil}
	root.Right.Left = &TreeNode{7, nil, nil}
	root.Right.Right = &TreeNode{9, nil, nil}
	root.Left.Left = &TreeNode{0, nil, nil}
	root.Left.Right = &TreeNode{4, nil, nil}
	root.Left.Right.Left = &TreeNode{3, nil, nil}
	root.Left.Right.Right = &TreeNode{5, nil, nil}
	q := root.Left
	p := root.Left.Right
	res := lowestCommonAncestor(root, p, q)
	fmt.Println(res.Val)
}

File: leetcode/tree/lowest_common_ancestor_bs.py
from common import TreeNode
from typing import List

class Solution:
    def __init__(self):
        self.ans = None
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def helper(curr):
            if not curr:
                return False
            left = helper(curr.left)
            right = helper(curr.right)
            mid = curr == p or curr == q
            if mid + left + right >= 2:
                self.ans = curr
            return mid or left or right
        helper(root)
        return self.ans

if __name__ == "__main__":
    root = TreeNode(3)
    root.left = TreeNode(5)
    root.right = TreeNode(1)
    root.left.left = TreeNode(6)
    root.left.right = TreeNode(2)
    root.left.right.left = TreeNode(7)
    root.left.right.right = TreeNode(4)
    root.right.left = TreeNode(0)
    root.right.right = TreeNode(8)
    p = root.left
    q = root.left.right.right
    res = Solution().lowestCommonAncestor(root, p, q)
    print(res.val)
